include/boost/url/impl/url_base.ipp

```diff
@@ -162,57 +162,101 @@ set_scheme_id(urls::scheme id)
-            return false;
-        if(impl_.nseg_ == 0)
-            return false;
-        BOOST_ASSERT(impl_.len(id_path) > 0);
-        if(s_[p] == '/')
-            return false;
-        if(! first_segment().contains(':'))
-            return false;
-        return true;
-    }();
-    if(! need_dot)
+    auto const po = impl_.offset(id_path);
+    bool const encode_colon =
+        !has_authority() &&
+        impl_.nseg_ > 0 &&
+        s_[po] != '/' &&
+        first_segment().contains(':');
+    if(!encode_colon)
     {
         // just remove the scheme
         resize_impl(id_scheme, 0, op);
         impl_.scheme_ = urls::scheme::none;
         check_invariants();
         return *this;
     }
-    // remove the scheme but add "./"
-    // to the beginning of the path
-    BOOST_ASSERT(n >= 2);
+    // encode any ":" in the first path segment
+    BOOST_ASSERT(sn >= 2);
+    auto pn = impl_.len(id_path);
+    std::size_t cn = 0;
+    for (char c: first_segment())
+        cn += c == ':';
+    std::size_t new_size =
+        size() - sn + 2 * cn;
+    bool need_resize = new_size > size();
+    if (need_resize)
+    {
+        resize_impl(
+            id_path, pn + 2 * cn, op);
+    }
     // move [id_scheme, id_path) left
     op.move(
         s_,
-        s_ + n,
-        p - n);
-    // move [id_path, id_end) left
+        s_ + sn,
+        po - sn);
+    // move [id_path, id_query) left
+    std::size_t qo = impl_.offset(id_query);
     op.move(
-        s_ + p - (n - 2),
-        s_ + p,
-        impl_.offset(id_end) - p);
+        s_ + po - sn,
+        s_ + po,
+        qo - po);
+    // move [id_query, id_end) left
+    op.move(
+        s_ + qo - sn + 2 * cn,
+        s_ + qo,
+        impl_.offset(id_end) - qo);
+
     // adjust part offsets.
-    // (p is invalidated)
-    impl_.adjust(id_user, id_path, 0-n);
-    impl_.adjust(id_query, id_end, 0-(n - 2));
-    auto dest = s_ + impl_.offset(id_path);
-    dest[0] = '.';
-    dest[1] = '/';
+    // (po and qo are invalidated)
+    if (need_resize)
+    {
+        impl_.adjust(id_user, id_end, 0 - sn);
+    }
+    else
+    {
+        impl_.adjust(id_user, id_path, 0 - sn);
+        impl_.adjust(id_query, id_end, 0 - sn + 2 * cn);
+    }
+    if (encode_colon)
+    {
+        // move the 2nd, 3rd, ... segments
+        auto begin = s_ + impl_.offset(id_path);
+        auto it = begin;
+        auto end = begin + pn;
+        while (*it != '/' &&
+               it != end)
+            ++it;
+        std::memmove(it + (2 * cn), it, end - it);
+
+        // move 1st segment
+        auto src = s_ + impl_.offset(id_path) + pn;
+        auto dest = s_ + impl_.offset(id_query);
+        src -= end - it;
+        dest -= end - it;
+        pn -= end - it;
+        do {
+            --src;
+            --dest;
+            if (*src != ':')
+            {
+                *dest = *src;
+            }
+            else
+            {
+                *dest-- = 'A';
```

Comment:
The user doesn't have a way to choose upper or lower case

Reply:
How would that happen? An extra parameter just in case this is required?

My initial reasoning they're supposed to choose, since normalization requires upper case and this is a form of normalization.

Reply:
we might not have a good way to let them pick it but I am just pointing it out. you should leave a comment